src/broadcast/broadcast.rs

```diff
@@ -449,9 +443,7 @@ impl<N: NodeIdT> Broadcast<N> {
             .skip(self.netinfo.num_correct() - self.netinfo.num_faulty() + self.fault_estimate)
             .take_while(|x| *x != self.our_id());
         for id in right {
-            if self.can_decodes.contains_key(p.root_hash())
-                && !self.can_decodes[p.root_hash()].contains(id)
-            {
+            if self.can_decodes.contains_key(hash) && !self.can_decodes[hash].contains(id) {
```

Comment:
That could also be done without an index access, and only a single map lookup:
```rust
if let Some(senders) = self.can_decodes.get(hash) {
    for id in right {
        if !senders.contains(id) {
```
Also, no reason to create `right` if `self.can_decodes.get(hash)` is `None`.
Alternatively, in a functional style:
```rust
if let Some(senders) = self.can_decodes.get(hash) {
    let right = …;
    let msgs = right.filter(|id| !senders.contains(id))
        .map(|id| Target::Node(id.clone()).message(echo_msg.clone()));
    step.messages.extend(msgs);
}
```